### PR TITLE
[Chk-1961] feat(wallet): manage services assignmet to wallet

### PR DESCRIPTION
**Pull request by ciuffagianluca**
Manage services assignment to the wallet

#### List of Changes

Implement the wallet service and wallet business logic to add and update the relationship between the wallet and the services. It use the LoggedAction paradigm to be sure that an event is stored in the collection when an operation is performed in a wallet

#### Motivation and Context

Wallet development to manage services

#### How Has This Been Tested?

Test has been performed locally and via junit tests

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/wallet/controllers/WalletController.kt [hunk 1]
```diff
@@ -51,17 +52,20 @@ class WalletController(
         walletId: UUID,
         exchange: ServerWebExchange?
     ): Mono<ResponseEntity<Void>> {
+        // TODO To be implemented
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/wallet/controllers/WalletController.kt:55]: what about making those service returning a custom exception to be map into [ExceptionHandler](https://github.com/pagopa/pagopa-wallet-service/blob/531e775fa8e96501a5feaab7e6ba3628935749eb/src/main/kotlin/it/pagopa/wallet/exceptionhandler/ExceptionHandler.kt#L25) to return a 501 http method?

File: src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt [hunk 2]
```diff
@@ -70,4 +76,126 @@ class WalletServiceTest {
 
         unmockkStatic(UUID::class)
     }
+
+    @Test
+    fun `should patch wallet document adding services`() {
```
Comment (pietro-tota) [inline src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt:81]:
```suggestion
    fun `should patch wallet document add services`() {
```

File: src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt [hunk 3]
```diff
@@ -70,4 +76,126 @@ class WalletServiceTest {
...
+        val walletArgumentCaptor: KArgumentCaptor<Wallet> = argumentCaptor<Wallet>()
+
+        given { walletRepository.findById(any<String>()) }
+            .willReturn(mono { WALLET_DOCUMENT_EMPTY_SERVICES_NULL_DETAILS_NO_PAYMENT_INSTRUMENT })
+
+        given { walletRepository.save(walletArgumentCaptor.capture()) }
+            .willReturn(mono { WALLET_DOCUMENT })
+
+        /* test */
+        assert(WALLET_DOMAIN_EMPTY_SERVICES_NULL_DETAILS_NO_PAYMENT_INSTRUMENT.services.isEmpty())
```
Comment (pietro-tota) [inline src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt:105]:
```suggestion
        assertTrue(WALLET_DOMAIN_EMPTY_SERVICES_NULL_DETAILS_NO_PAYMENT_INSTRUMENT.services.isEmpty())
```

File: src/main/kotlin/it/pagopa/wallet/controllers/WalletController.kt (hunk 1 above)
Comment (giovanniberti) [reply src/main/kotlin/it/pagopa/wallet/controllers/WalletController.kt:55]: That is out of scope for the PR, for now the aim is to lay the groundwork of domain modeling + auditing, the true bussiness logic implementation will be done in the next sprint 👌

**Review by pietro-tota — DISMISSED** [review]

File: src/main/kotlin/it/pagopa/wallet/services/WalletService.kt [hunk 4]
```diff
@@ -41,4 +46,45 @@ class WalletService(@Autowired private val walletRepository: WalletRepository) {
...
+            .flatMap { walletRepository.save(it) }
+            .map { LoggedAction(it.toDomain(), WalletPatchEvent(it.id)) }
+    }
+
+    private fun updateServiceList(
+        wallet: it.pagopa.wallet.documents.wallets.Wallet,
+        service: Pair<ServiceName, ServiceStatus>
+    ): it.pagopa.wallet.documents.wallets.Wallet {
+        val updatedServiceList = wallet.services.toMutableList()
+        when (val index = wallet.services.indexOfFirst { s -> s.name == service.first.name }) {
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/wallet/services/WalletService.kt:67]: Are we sure that we want to add a service when patching if the service is not present?

File: src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt [hunk 5]
```diff
@@ -70,4 +77,128 @@ class WalletServiceTest {
...
+            )
+            .expectNextMatches {
+                expectedLoggedAction.data.id == WALLET_UUID &&
+                    expectedLoggedAction.data.userId == USER_ID &&
+                    expectedLoggedAction.data.services.isEmpty() &&
+                    expectedLoggedAction.data.status == WalletStatusDto.CREATED &&
+                    expectedLoggedAction.data.contractId == CONTRACT_ID &&
+                    expectedLoggedAction.data.details == null &&
+                    expectedLoggedAction.data.paymentInstrumentId == null &&
+                    expectedLoggedAction.data.paymentMethodId == PAYMENT_METHOD_ID
```
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt]: We have a problem mocking the `instant.now()`. So I use the expectNextMatches as workaround

File: src/main/kotlin/it/pagopa/wallet/services/WalletService.kt (hunk 4 above)
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/wallet/services/WalletService.kt:67]: Good point. I'm not really sure about this, but, think about the use case when a new service is added to the service repo and a single user want to add it to his wallet. In this case we need this logic. Anyway we could think about having different logic for this 2 usecase, while They are logically similar from the point of view of "patch the wallet" action.

File: src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt (hunk 2 above)
Comment (giovanniberti) [reply src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt]: Fixed in a42f5e8

File: src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (giovanniberti) [reply src/test/kotlin/it/pagopa/wallet/services/WalletServiceTest.kt]: Fixed in 7e540ba

**Review by pietro-tota — APPROVED** [review]
🚀

**Pull request merged**